drivers/network/tcpip/tcpip/icmp.c

```diff
@@ -0,0 +1,389 @@
+        PUCHAR CurrentUserBuffer;
+
+        TI_DbgPrint(DEBUG_ICMP, ("RemainingSize: %u, RoundTripTime: %u\n", Context->RemainingSize, RoundTripTime));
+
+        if (Context->RemainingSize >= sizeof(ICMP_ECHO_REPLY))
+        {
+            memcpy(&CurrentReply->Address, SourceAddress, sizeof(CurrentReply->Address));
+            CurrentReply->Status = GetReplyStatus(IcmpHeader);
+            CurrentReply->RoundTripTime = RoundTripTime;
+            CurrentReply->Reserved = 0;
+            CurrentReply->Data = NULL;
+            CurrentReply->DataSize = 0;
+            CurrentReply->Options.Ttl = Header->Ttl;
+            CurrentReply->Options.Tos = Header->Tos;
+            CurrentReply->Options.Flags = Header->FlagsFragOfs >> 13;
+            CurrentReply->Options.OptionsData = NULL;
+            CurrentReply->Options.OptionsSize = 0;
+
+            Context->RemainingSize -= sizeof(ICMP_ECHO_REPLY);
+            Context->CurrentReply += sizeof(ICMP_ECHO_REPLY);
+        }
+
+        CurrentUserBuffer = (PUCHAR)Context->Irp->UserBuffer + (Context->CurrentReply - (PUCHAR)Context->Irp->AssociatedIrp.SystemBuffer);
+
+        if (DataSize > 0 && Context->RemainingSize > 0)
+        {
+            UINT32 _DataSize = min(Context->RemainingSize, DataSize);
+
+            memcpy(Context->CurrentReply + Context->RemainingSize - _DataSize, DataBuffer, _DataSize);
+            CurrentReply->Data = CurrentUserBuffer + Context->RemainingSize - _DataSize;
+            CurrentReply->DataSize = _DataSize;
+
+            Context->RemainingSize -= _DataSize;
+            // Context->ReplyBuffer += _DataSize;
+        }
+        else
+        {
+            TI_DbgPrint(DEBUG_ICMP, ("RemainingSize: %u, DataSize: %d\n", Context->RemainingSize, DataSize));
+        }
+
+        if (OptionsLength > 0 && Context->RemainingSize > 0)
+        {
+            UINT32 _OptSize = min(Context->RemainingSize, OptionsLength);
+
+            memcpy(Context->CurrentReply + Context->RemainingSize + _OptSize, Options, _OptSize);
+            CurrentReply->Options.OptionsData = CurrentUserBuffer + Context->RemainingSize + _OptSize;
+            CurrentReply->Options.OptionsSize = _OptSize;
+
+            Context->RemainingSize -= _OptSize;
+            // Context->ReplyBuffer += _OptSize;
+        }
+        else
+        {
+            TI_DbgPrint(DEBUG_ICMP, ("RemainingSize: %u, OptSize: %d\n", Context->RemainingSize, OptionsLength));
+        }
+
+        Context->nReplies++;
+    }
+
+    if (NT_ERROR(Context->ErrorStatus) || Context->RemainingSize < sizeof(ICMP_ECHO_REPLY))
+    {
+        TI_DbgPrint(DEBUG_ICMP, ("The space is over: %u, or error happened: 0x%x\n",
+            Context->RemainingSize, Context->ErrorStatus));
+
+        // if the timer was inserted, that means DPC has not been queued yet
+        if (KeCancelTimer(&TimeoutTimer))
+        {
+            ASSERT(KeInsertQueueDpc(&TimeoutDpc, NULL, NULL));
```

Comment:
I have rewritten this code several times. From [this](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-kecanceltimer) article:
> The driver can check the return code of KeCancelTimer to determine if the DPC is running.
I've checked the code of `KiCompleteTimer` and `KeCancelTimer` and as far as I can see, if `KiCompleteTimer` is called before `KeCancelTimer`, the locks will do the work and `KeCancelTimer` will wail for completion and return false.

What can be the other way to avoid race conditions between timer routine and receive routine?